swarm/network/stream/delivery.go

```diff
@@ -68,46 +64,39 @@ type RetrieveRequestMsg struct {
 }
 
 func (d *Delivery) handleRetrieveRequestMsg(ctx context.Context, sp *Peer, req *RetrieveRequestMsg) error {
-	log.Trace("received request", "peer", sp.ID(), "hash", req.Addr)
+	log.Trace("handle retrieve request", "peer", sp.ID(), "hash", req.Addr)
 	handleRetrieveRequestMsgCount.Inc(1)
 
-	var osp opentracing.Span
-	ctx, osp = spancontext.StartSpan(
+	ctx, osp := spancontext.StartSpan(
 		ctx,
-		"stream.handle.retrieve")
+		"handle.retrieve.request")
 
 	osp.LogFields(olog.String("ref", req.Addr.String()))
 
-	var cancel func()
-	// TODO: do something with this hardcoded timeout, maybe use TTL in the future
-	ctx = context.WithValue(ctx, "peer", sp.ID().String())
-	ctx = context.WithValue(ctx, "hopcount", req.HopCount)
-	ctx, cancel = context.WithTimeout(ctx, network.RequestTimeout)
+	defer osp.Finish()
 
-	go func() {
-		select {
-		case <-ctx.Done():
-		case <-d.quit:
-		}
-		cancel()
-	}()
-
-	go func() {
-		defer osp.Finish()
-		ch, err := d.netStore.Get(ctx, chunk.ModeGetRequest, req.Addr)
-		if err != nil {
-			retrieveChunkFail.Inc(1)
-			log.Debug("ChunkStore.Get can not retrieve chunk", "peer", sp.ID().String(), "addr", req.Addr, "hopcount", req.HopCount, "err", err)
-			return
-		}
-		syncing := false
+	ctx, cancel := context.WithTimeout(ctx, timeouts.FetcherGlobalTimeout)
+	defer cancel()
 
-		err = sp.Deliver(ctx, ch, Top, syncing)
-		if err != nil {
-			log.Warn("ERROR in handleRetrieveRequestMsg", "err", err)
-		}
-		osp.LogFields(olog.Bool("delivered", true))
-	}()
+	r := &storage.Request{
+		Addr:     req.Addr,
+		Origin:   sp.ID(),
+		HopCount: req.HopCount,
+	}
+	chunk, err := d.netStore.Get(ctx, chunk.ModeGetRequest, r)
+	if err != nil {
+		retrieveChunkFail.Inc(1)
+		log.Debug("ChunkStore.Get can not retrieve chunk", "peer", sp.ID().String(), "addr", req.Addr, "hopcount", req.HopCount, "err", err)
+		return nil
+	}
+
+	log.Trace("retrieve request, delivery", "ref", req.Addr, "peer", sp.ID())
+	syncing := false
+	err = sp.Deliver(ctx, chunk, syncing)
+	if err != nil {
+		log.Warn("ERROR in handleRetrieveRequestMsg", "err", err)
```

Comment:
if it's an error maybe we should change to `log.Error`? how often does this happen?

Reply:
yes, this is def an error, and not a warning.